engine/runtime/src/main/java/org/enso/interpreter/node/controlflow/caseexpr/TimeZoneBranchNode.java

```diff
@@ -0,0 +1,59 @@
+package org.enso.interpreter.node.controlflow.caseexpr;
+
+import com.oracle.truffle.api.RootCallTarget;
+import com.oracle.truffle.api.dsl.Fallback;
+import com.oracle.truffle.api.dsl.Specialization;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.interop.InteropLibrary;
+import com.oracle.truffle.api.library.CachedLibrary;
+import com.oracle.truffle.api.nodes.NodeInfo;
+import com.oracle.truffle.api.profiles.ConditionProfile;
+import org.enso.interpreter.runtime.data.EnsoTimeZone;
+import org.enso.interpreter.runtime.data.Type;
+
+/** An implementation of the case expression specialised to working on Time_Zone. */
+@NodeInfo(shortName = "DateMatch")
+public abstract class TimeZoneBranchNode extends BranchNode {
+  private final Type timeZone;
+  private final ConditionProfile profile = ConditionProfile.createCountingProfile();
+
+  TimeZoneBranchNode(Type vector, RootCallTarget branch) {
+    super(branch);
+    this.timeZone = vector;
+  }
+
+  /**
+   * Creates a new node for handling matching on a case expression.
+   *
+   * @param timeZone the expression to use for matching
+   * @param branch the expression to be executed if (@code matcher} matches
+   * @return a node for matching in a case expression
+   */
+  public static TimeZoneBranchNode build(Type timeZone, RootCallTarget branch) {
+    return TimeZoneBranchNodeGen.create(timeZone, branch);
+  }
+
+  @Specialization
+  void doType(VirtualFrame frame, Object state, Type target) {
+    if (profile.profile(timeZone == target)) {
+      accept(frame, state, new Object[0]);
+    }
+  }
+
+  @Specialization
+  void doEnsoZone(VirtualFrame frame, Object state, EnsoTimeZone timeZone) {
+    accept(frame, state, new Object[0]);
+  }
+
+  @Specialization(guards = "interop.isTimeZone(zone)")
```

Comment:
Here also, I understand that our `ZonedDateTime` will answer true to `isTimeZone`, but _really_ I don't think the `Time_Zone` branch should match it, right?

Reply:
Yeah, we had that discussion with @JaroslavTulach. It's a bit of a pickle because I think it should match because it kind of _inherits_ time zone interface (not literally but I think you know what I mean).

Reply:
I get it but I wouldn't say it inherits it.

It _does_ expose the common interface - yes.

But I'd say that a Date_Time _contains_ a time-zone rather than _is_ a time-zone. It is a richer object consisting of multiple components. 

I think it boils down to - do I want to be able to use a `Date_Time` wherever a `Time_Zone` is expected? In my opinion - no - not at all - as this kind of implicit conversion will likely just cause bugs. If I want to get the timezone I'd call `datetime.zone`, same for `datetime.date : Date` if I need a `Date` instead of automatically treating a `Date_Time` as a `Date`.

That is just my opinion though, I think it would be good to discuss what we want this to work like from the user's perspective. @jdunkerley do you have any thoughst on it?

I think if we make very clear that `Date_Time <: Date` and `Date_Time <: Time_Zone`, then we can do it. I just envision a users writing code like:
```
case x of
    Date -> ...
    Date_Time -> ...
```
and being confused why the second branch is never chosen. So for the sake of simplicity I'd avoid introducing these subtyping relationships. But if we really want to, I think we can do, if we document this well enough.